{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "Loading required package: parallel\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:parallel’:\n",
      "\n",
      "    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,\n",
      "    clusterExport, clusterMap, parApply, parCapply, parLapply,\n",
      "    parLapplyLB, parRapply, parSapply, parSapplyLB\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: scuttle\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(\"splatter\")\n",
    "library(\"scater\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "generate_sim <- function(n_ctypes,ctype_de_prob,n_donors,cells_per_don,\n",
    "                         n_genes,n_process,de_strength,condition.prob,\n",
    "                         similarity.scale, sim_seed=1) {\n",
    "  coldat_lst <- list()\n",
    "  rowdat_lst <- list()\n",
    "  sim_objs <- list()\n",
    "  \n",
    "  vcf <- mockVCF(n.samples = n_donors)\n",
    "  gff <- mockGFF(n.genes = n_genes)\n",
    "  \n",
    "  for (i in 1:n_process) {\n",
    "    seed <- i * sim_seed\n",
    "    \n",
    "    # make sure group.prob sums to 1\n",
    "    gpr <- rep(1/n_ctypes, n_ctypes)\n",
    "    gpr <- round(gpr,2)\n",
    "    gpr[length(gpr)] <- 1 - sum(gpr[1:(length(gpr)-1)])\n",
    "    \n",
    "    # now trying for multi population two condition sim and adding more donors\n",
    "    params.cond <- newSplatPopParams(similarity.scale = similarity.scale,\n",
    "                                     condition.prob = condition.prob,\n",
    "                                     cde.prob = c(0,.1),\n",
    "                                     cde.downProb = c(0,.5),\n",
    "                                     cde.facLoc = de_strength, \n",
    "                                     cde.facScale = 0.5,\n",
    "                                     de.prob = ctype_de_prob,\n",
    "                                     de.facLoc = 0.5, \n",
    "                                     de.facScale = 0.5,\n",
    "                                     group.prob = gpr,\n",
    "                                     batchCells = c(cells_per_don),\n",
    "                                     batch.size = n_donors, nGenes=n_genes,\n",
    "                                     eqtl.n = 0,\n",
    "                                     seed = seed)\n",
    "    \n",
    "    sim.pop.cond1 <- splatPopSimulate(vcf = vcf, gff = gff,\n",
    "                                      params = params.cond, \n",
    "                                      sparsify = FALSE)\n",
    "    \n",
    "    # need to make cell names unique per donor and group\n",
    "    colnames(sim.pop.cond1) <- paste(sim.pop.cond1$Sample, sim.pop.cond1$Group, sim.pop.cond1$Cell, sep=\"_\")\n",
    "    \n",
    "    sim.pop.cond1 <- logNormCounts(sim.pop.cond1)\n",
    "    simcounts1 <- counts(sim.pop.cond1)\n",
    "    simcounts1 <- methods::as(as.matrix(simcounts1),'sparseMatrix')\n",
    "    simmeta1 <- as.data.frame(colData(sim.pop.cond1)@listData)\n",
    "    colnames(simmeta1)[4] <- 'donors'\n",
    "    colnames(simmeta1)[5] <- 'ctypes'\n",
    "    \n",
    "    coldat_lst[[i]] <- simmeta1\n",
    "    rowdat_lst[[i]] <- rowData(sim.pop.cond1)\n",
    "    sim_objs[[i]] <- sim.pop.cond1\n",
    "  }\n",
    "  \n",
    "  new_simcounts <- combine_processes(sim_objs,coldat_lst,rowdat_lst)\n",
    "  \n",
    "  return(list(new_simcounts,coldat_lst,rowdat_lst,sim_objs))\n",
    "}\n",
    "\n",
    "\n",
    "combine_processes <- function(sim_objs,coldat_lst,rowdat_lst) {\n",
    "  simmeta1 <- coldat_lst[[1]]\n",
    "  sim.pop.cond1 <- sim_objs[[1]]\n",
    "  d_list <- unique(simmeta1$donors)\n",
    "  ctypes <- unique(simmeta1$ctypes)\n",
    "  \n",
    "  # matrix to store new combined data\n",
    "  new_log_norm <- matrix(ncol=ncol(logcounts(sim.pop.cond1)), nrow=nrow(logcounts(sim.pop.cond1)))\n",
    "  colnames(new_log_norm) <- colnames(logcounts(sim.pop.cond1))\n",
    "  rownames(new_log_norm) <- rownames(logcounts(sim.pop.cond1))\n",
    "  for (dnr in d_list) {\n",
    "    for (ct in ctypes) {\n",
    "      d_counts_list <- list()\n",
    "      for (pr_ndx in 1:length(sim_objs)) {\n",
    "        simmeta1 <- coldat_lst[[pr_ndx]]\n",
    "        sim.pop.cond1 <- sim_objs[[pr_ndx]]\n",
    "        meta_sub1 <- simmeta1[simmeta1$donors==dnr,]\n",
    "        d_cells1 <- rownames(meta_sub1)[meta_sub1$ctypes==ct]\n",
    "        d_counts1 <- logcounts(sim.pop.cond1)[,d_cells1]\n",
    "        \n",
    "        d_counts_list[[pr_ndx]] <- d_counts1\n",
    "      }\n",
    "      \n",
    "      # compute average expression\n",
    "      running_total <- d_counts_list[[1]]\n",
    "      for (pr_ndx in 2:length(d_counts_list)) {\n",
    "        running_total <- running_total + d_counts_list[[pr_ndx]]\n",
    "      }\n",
    "      d_counts_av <- running_total / length(d_counts_list)\n",
    "      \n",
    "      # store results\n",
    "      new_log_norm[rownames(d_counts_av),colnames(d_counts_av)] <- d_counts_av\n",
    "    }\n",
    "  }\n",
    "  \n",
    "  # exponentiate the data\n",
    "  new_log_norm2 <- 2**new_log_norm\n",
    "  \n",
    "  # subtract 1\n",
    "  new_log_norm2 <- new_log_norm2 - 1\n",
    "  \n",
    "  # divide by scale factor\n",
    "  lib_sizes <- colSums(counts(sim_objs[[1]])) # arbitrarily using the first sim for lib sizes\n",
    "  scale_factor <- mean(lib_sizes)\n",
    "  new_log_norm2 <- new_log_norm2 / scale_factor\n",
    "  \n",
    "  # adjust fractional counts to sum to 1\n",
    "  adjustments <- 1/colSums(new_log_norm2)\n",
    "  new_log_norm2 <- sweep(new_log_norm2,MARGIN=2,adjustments,FUN='*')\n",
    "  \n",
    "  # multiply to get correct original sizes\n",
    "  new_log_norm2 <- sweep(new_log_norm2,MARGIN=2,lib_sizes,FUN='*')\n",
    "  ## these are honestly too large... should add sparsity or adjust the param\n",
    "  \n",
    "  new_simcounts <- round(new_log_norm2)\n",
    "  new_simcounts <- methods::as(as.matrix(new_simcounts),'sparseMatrix')\n",
    "  \n",
    "  return(new_simcounts)\n",
    "}\n",
    "\n",
    "get_sim_auc <- function(container, coldat_lst, rowdat_lst) {\n",
    "  all_auc <- c()\n",
    "  all_pv <- c()\n",
    "  for (i in 1:length(coldat_lst)) {\n",
    "    # determine donors of condition 1 for the given process\n",
    "    coldat <- coldat_lst[[i]]\n",
    "    dg1 <- unique(coldat$donors[coldat$Condition=='Condition1'])\n",
    "    \n",
    "    # determine which factor best distinguishes these donors\n",
    "    dscores <- container$tucker_results[[1]]\n",
    "    fact_links <- abs(colMeans(dscores[dg1,]))\n",
    "    ndx_max <- order(fact_links,decreasing=TRUE)[1]\n",
    "    \n",
    "    # evaluate AUC\n",
    "    # # do factor 1 first\n",
    "    # sig_vectors <- get_significance_vectors(container,\n",
    "    #                                         factor_select=ndx_max, container$experiment_params$ctypes_use)\n",
    "    # # convert list to df\n",
    "    # sig_df <- t(as.data.frame(do.call(rbind, sig_vectors)))\n",
    "    # pred1 <- c(sig_df)\n",
    "    \n",
    "    ## trying using loadings instead\n",
    "    f_res <- get_one_factor(container,ndx_max)\n",
    "    sig_df <- f_res[[2]]\n",
    "    pred1 <- abs(c(sig_df))\n",
    "    ##\n",
    "    \n",
    "    # should make this work with more than two cell types\n",
    "    rowdat <- rowdat_lst[[i]]\n",
    "    ct1_de <- rowdat[rownames(sig_df),c('ConditionDE.Condition1','ConditionDE.Condition2')]\n",
    "    ct1_de <- as.matrix(ct1_de)\n",
    "    ct1_de_logical <- rowSums(ct1_de!=1)>0\n",
    "\n",
    "    ## remove exceptions here\n",
    "    \n",
    "    ##\n",
    "    \n",
    "    # de_true1 <- c(ct1_de_logical1,ct2_de_logical)\n",
    "    de_true1 <- rep(ct1_de_logical,ncol(sig_df)) # repeated for each ctype\n",
    "    \n",
    "    pROC_obj <- roc(de_true1,pred1,\n",
    "                    smoothed = TRUE,\n",
    "                    plot=FALSE, AUC=TRUE)\n",
    "    auc1 <- pROC_obj[[\"auc\"]]\n",
    "    # print(auc1)\n",
    "    \n",
    "    # store AUC result\n",
    "    all_auc <- c(all_auc,auc1)\n",
    "    \n",
    "    ## now compute donor process association p-values\n",
    "    coldat <- unique(coldat[,c('donors','Condition')])\n",
    "    rownames(coldat) <- coldat$donors\n",
    "    dsc <- f_res[[1]]\n",
    "    tmp <- cbind.data.frame(dsc,coldat[rownames(dsc),'Condition'])\n",
    "    colnames(tmp) <- c('dscore','Condition')\n",
    "    lmres <- summary(lm(as.formula('dscore~Condition'),data=tmp))\n",
    "    pval <- stats::pf(lmres$fstatistic[1],lmres$fstatistic[2],lmres$fstatistic[3],lower.tail=FALSE)\n",
    "    all_pv <- c(all_pv,-log10(pval))\n",
    "  }\n",
    "  return(list(all_auc,all_pv))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "get_performance_range <- function(param_test, param_range, n_iter, rand_seeds,\n",
    "                                   n_ctypes=2,ctype_de_prob=.3,n_donors=50,\n",
    "                                   cells_per_don=50,n_genes=2000,n_process=2,\n",
    "                                   de_strength=1,condition.prob=c(.25,.75),\n",
    "                                   similarity.scale=1.5) {\n",
    "  \n",
    "  # matrix to store results\n",
    "  myres <- matrix(ncol=4,nrow=0)\n",
    "  colnames(myres) <- c('mean_auc','mean_pv','method','param_val')\n",
    "  myres_summary <- matrix(ncol=6,nrow=0)\n",
    "  colnames(myres_summary) <- c('mean_auc','sd_auc','mean_pv','sd_pv','method','param_val')\n",
    "  \n",
    "  for (i in 1:length(param_range)) {\n",
    "    \n",
    "    # need to switch parameter value for param_test\n",
    "    if (param_test=='n_ctypes') {\n",
    "      n_ctypes <- param_range[i]\n",
    "    } else if (param_test=='ctype_de_prob') {\n",
    "      ctype_de_prob <- param_range[i]\n",
    "    } else if (param_test=='n_donors') {\n",
    "      n_donors <- param_range[i]\n",
    "    } else if (param_test=='cells_per_don') {\n",
    "      cells_per_don <- param_range[i]\n",
    "    } else if (param_test=='n_genes') {\n",
    "      n_genes <- param_range[i]\n",
    "    } else if (param_test=='n_process') {\n",
    "      n_process <- param_range[i]\n",
    "    } else if (param_test=='de_strength') {\n",
    "      de_strength <- param_range[i]\n",
    "    } else if (param_test=='similarity.scale') {\n",
    "      similarity.scale <- param_range[i]\n",
    "    } else {\n",
    "      stop('choose one of the available parameter options')\n",
    "    }\n",
    "    \n",
    "    for (j in 1:n_iter) {\n",
    "      print(i)\n",
    "      print(j)\n",
    "      sim_res <- generate_sim(n_ctypes=n_ctypes,ctype_de_prob=ctype_de_prob,\n",
    "                              n_donors=n_donors,\n",
    "                              cells_per_don=cells_per_don,n_genes=n_genes,n_process=n_process,\n",
    "                              de_strength=de_strength,condition.prob=condition.prob,\n",
    "                              similarity.scale=similarity.scale, sim_seed=(rand_seeds[i]+j))\n",
    "      \n",
    "      new_simcounts <- sim_res[[1]]\n",
    "      coldat_lst <- sim_res[[2]]\n",
    "      rowdat_lst <- sim_res[[3]]\n",
    "      sim_objs <- sim_res[[4]]\n",
    "      \n",
    "      ctypes_use <- sapply(1:n_ctypes,function(x){\n",
    "        paste0('Group',x)\n",
    "      })\n",
    "      \n",
    "      param_list <- initialize_params(ctypes_use = ctypes_use,\n",
    "                                      ncores = 30, rand_seed = 10)\n",
    "      \n",
    "      sim_container <- make_new_container(count_data=new_simcounts, meta_data=coldat_lst[[1]], params=param_list)\n",
    "      \n",
    "      sim_container <- form_tensor(sim_container, donor_min_cells=3,\n",
    "                                   norm_method='trim', scale_factor=10000,\n",
    "                                   vargenes_method='norm_var_pvals', vargenes_thresh=1,\n",
    "                                   scale_var = TRUE, var_scale_power = .5)\n",
    "\n",
    "      # get AUC and pvals with Tucker\n",
    "      sim_container <- run_tucker_ica(sim_container, ranks=c(n_process,n_process*2),\n",
    "                                      tucker_type = 'regular', rotation_type = 'hybrid')\n",
    "      \n",
    "      all_tucker_res <- get_sim_auc(sim_container,coldat_lst,rowdat_lst)\n",
    "      tucker_auc <- mean(all_tucker_res[[1]])\n",
    "      tucker_pv <- mean(all_tucker_res[[2]])\n",
    "      row_add <- data.frame('mean_auc'=tucker_auc,'mean_pv'=tucker_pv,'method'='Tucker','param_val'=param_range[i])\n",
    "      myres <- rbind.data.frame(myres,row_add) # store results\n",
    "      \n",
    "      # get AUC and pvals with PCA\n",
    "      pca_unfolded(sim_container,n_process)\n",
    "      all_pca_res <- get_sim_auc(sim_container,coldat_lst,rowdat_lst)\n",
    "      pca_auc <- mean(all_pca_res[[1]])\n",
    "      pca_pv <- mean(all_pca_res[[2]])\n",
    "      row_add <- data.frame('mean_auc'=pca_auc,'mean_pv'=pca_pv,'method'='PCA','param_val'=param_range[i])\n",
    "      myres <- rbind.data.frame(myres,row_add) # store results\n",
    "      \n",
    "      # get AUC and pvals with NMF\n",
    "      nmf_unfolded(sim_container,n_process)\n",
    "      all_nmf_res <- get_sim_auc(sim_container,coldat_lst,rowdat_lst)\n",
    "      nmf_auc <- mean(all_nmf_res[[1]])\n",
    "      nmf_pv <- mean(all_nmf_res[[2]])\n",
    "      row_add <- data.frame('mean_auc'=nmf_auc,'mean_pv'=nmf_pv,'method'='NMF','param_val'=param_range[i])\n",
    "      myres <- rbind.data.frame(myres,row_add) # store results\n",
    "    }\n",
    "    myres_sub <- myres[myres$param_val==param_range[i],]\n",
    "    # auc_summary = tapply(myres_sub$mean_auc, myres_sub$method, mean) \n",
    "    # pv_summary = tapply(myres_sub$mean_pv, myres_sub$method, mean) \n",
    "    # pv_summary <- pv_summary[names(auc_summary)] # ordering them same way\n",
    "    \n",
    "    auc_mean = tapply(myres_sub$mean_auc, myres_sub$method, mean) \n",
    "    pv_mean = tapply(myres_sub$mean_pv, myres_sub$method, mean) \n",
    "    pv_mean <- pv_mean[names(auc_mean)] # ordering them same way\n",
    "    \n",
    "    auc_sd = tapply(myres_sub$mean_auc, myres_sub$method, sd) \n",
    "    pv_sd = tapply(myres_sub$mean_pv, myres_sub$method, sd) \n",
    "    pv_sd <- pv_sd[names(auc_sd)] # ordering them same way\n",
    "    \n",
    "    row_add <- data.frame('mean_auc'=auc_mean,'sd_auc'=auc_sd,'mean_pv'=pv_mean,'sd_pv'=pv_sd,'method'=names(pv_mean),'param_val'=param_range[i])\n",
    "    myres_summary <- rbind.data.frame(myres_summary,row_add)\n",
    "  }\n",
    "  return(list(myres,myres_summary))\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 75055 14928 47323 55865\n"
     ]
    }
   ],
   "source": [
    "r_seeds <- round(runif(4, min=0, max=100000)) # should be length of param_range below\n",
    "print(r_seeds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_res <- get_performance_range(param_test='n_process', param_range=c(2,3,4,5), n_iter=10, \n",
    "                      rand_seeds=r_seeds, n_ctypes=2, ctype_de_prob=.3, n_donors=50,\n",
    "                      cells_per_don=25, n_genes=500, n_process=2,\n",
    "                      de_strength=1, condition.prob=c(.25,.75),\n",
    "                      similarity.scale=1.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "saveRDS(all_res,file='/home/jmitchel/data/sim_data/iter_n_proc.rds')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] 9\n"
     ]
    }
   ],
   "source": [
    "print(9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
